Modules/Quotas/src/Quotas.Application/Extensions/IdentityMetricExtensions.cs

```diff
@@ -0,0 +1,60 @@
+﻿using Backbone.Modules.Quotas.Application.Infrastructure.Persistence.Repository;
+using Backbone.Modules.Quotas.Domain;
+using Backbone.Modules.Quotas.Domain.Aggregates.Identities;
+using Enmeshed.BuildingBlocks.Domain;
+using Enmeshed.Tooling;
+
+namespace Backbone.Modules.Quotas.Application.Extensions;
+public static class IdentityMetricExtensions
+{
+    public static async Task UpdateMetrics(this Identity identity, IEnumerable<string> metrics, IMetricCalculatorFactory factory, IQuotasRepository quotasRepository, CancellationToken cancellationToken)
+    {
+        foreach (var metric in metrics)
+        {
+            var metricCalculator = factory.CreateFor(metric);
+            await identity.UpdateMetricAsync(metric, metricCalculator, quotasRepository, cancellationToken);
+        }
+        return;
+    }
+
+    private static async Task UpdateMetricAsync(this Identity identity, string metric, IMetricCalculator metricCalculator, IQuotasRepository quotasRepository, CancellationToken cancellationToken)
+    {
+        var quotas = await GetAppliedQuotasForMetricAsync(metric, quotasRepository, cancellationToken);
+        var unExhaustedQuotas = quotas.Where(q => q.IsExhaustedUntil is null || q.IsExhaustedUntil > SystemTime.UtcNow);
+        foreach (var quota in unExhaustedQuotas)
+        {
+            var newValue = metricCalculator.CalculateUsage(
+                SystemTime.UtcNow.AddDays(-30),
+                SystemTime.UtcNow,
+                identity.Address);
+
+            quota.Update(newValue);
+        }
+    }
+
+    private static async Task<IEnumerable<Quota>> GetAppliedQuotasForMetricAsync(string metric, IQuotasRepository quotasRepository, CancellationToken cancellationToken)
+    {
+        var allQuotasOfMetric = await quotasRepository.FindQuotasByMetricKey(metric, cancellationToken);
+        var highestWeight = allQuotasOfMetric.OrderByDescending(q => q.Weight).FirstOrDefault().Weight;
+        var appliedQuotas = allQuotasOfMetric.Where(q => q.Weight == highestWeight && q.isCurrentlyValid()).ToList();
+        return appliedQuotas;
+    }
+
+    private static void Update(this Quota quota, uint newValue)
+    {
+        var periodEndTime = SystemTime.UtcNow; // where should this come from?
+        var newIsExhaustedUntil = SystemTime.UtcNow; // where should this come from?
+        DateTime? IsExhaustedUntil = SystemTime.UtcNow;
+        if (newValue >= quota.Max)
+        {
+            if(newIsExhaustedUntil == null || periodEndTime > newIsExhaustedUntil)
+            {
+                IsExhaustedUntil = periodEndTime; // where to set this?
+            }
+        }
+        else
+        {
+            IsExhaustedUntil = null;
+        }
+    }
```

Comment:
The sample for this method is rather confusing in regards to what comes from where.
It might be useful for us to review the logic behind it.

Reply:
First of all: this is domain logic. So don't make an extension method out of it.

Regarding the period end time: it can be calculated from the period of the quota. So for example, if I would execute this method on June 12 2023:
- if the period is "Day", then periodEndTime would be 2023-06-12T23:59:59.9999
- if the period is "Month", then periodEndTime would be 2023-06-30T23:59:59.9999
- if the period is "Year", then periodEndTime would be 2023-12-31T23:59:59.9999
- ...

The variable `newIsExhaustedUntil` isn't needed. You can set directly set the `IsExhaustedUntil` property of the Quota. The value depends on whether `newValue` is greater than or equal to `quota.Max`. If it is, then the new exhaustion time is the period end time. If it isn't then it should be set to null.

Reply:
> First of all: this is domain logic. So don't make an extension method out of it.

Should we put this login in a Mediatr command?

Reply:
Since it's domain logic, it should be located in the Domain project (-> the Quota class)